### PR TITLE
Fix all warnings and add -Xfatal-warnings

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Needs https://github.com/lampepfl/scala/pull/28, which removes warnings in the backend

File: project/Build.scala [hunk 1]
```diff
@@ -574,6 +574,11 @@ object Build {
       // Add git-hash used to package the distribution to the manifest to know it in runtime and report it in REPL
       packageOptions += ManifestAttributes(("Git-Hash", VersionUtil.gitHash)),
 
+      scalacOptions ++= Seq(
```
Comment (allanrenucci) [inline project/Build.scala]: Should this go into `commonSettings`?

File: compiler/src/dotty/tools/dotc/core/Symbols.scala [hunk 2]
```diff
@@ -722,23 +722,23 @@ object Symbols {
     def clear(): Unit = value.clear()
 
     def filter(p: ((Symbol, T)) => Boolean): Map[Symbol, T] = {
-      import scala.collection.JavaConversions._
-      value.toMap.filter(p)
+      import scala.collection.JavaConverters._
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Symbols.scala]: I would just move the import at the top of the file.

File: project/Build.scala [hunk 3]
```diff
@@ -574,6 +574,11 @@ object Build {
...
+        "-Xfatal-warnings",
```
Comment (smarter) [inline project/Build.scala]: I would enable this on all projects but only on the CI, when doing local development warnings as errors is just annoying.

Comment (nicolasstucki) [reply project/Build.scala]: Good idea

File: project/Build.scala (hunk 1 above)
Comment (nicolasstucki) [reply project/Build.scala]: Cannot do it due to #3543

Comment (nicolasstucki) [reply project/Build.scala]: We should push that for another PR

**Comment by nicolasstucki** [comment]
Rebased on top of #3545 (last warning fix).

**Comment by nicolasstucki** [comment]
@allanrenucci, I rebased it and fixed a new warning that came from some recently merged PR

**Comment by allanrenucci** [comment]
@nicolasstucki Can you open an issue for the unchecked reference warning?

**Pull request merged**